### PR TITLE
POC: Move testdata datasource to a new separate structure

### PR DESCRIPTION
**Pull request by marefr**
Goal of this POC is to try and move the testdata datasource frontend/backend code to a new directory where it uses the same structure as an external plugin. If possible, trying to keep using some Grafana internals as dependencies for now to limit the scope of changes.

Ref #68973

**State of things so far:**
- :heavy_check_mark:  Defining a Go sub module together with using Go workspaces seems to work and allow the plugins/grafana-test-datasource to import code from ./pkg
- :heavy_check_mark: Grafana core now successfully imports dependencies from ./public/plugins/grafana-test-datasource/pkg/plugin as Go module and the old backend testdatasource directory is all gone
- :heavy_check_mark:  ./public/plugins/grafana-test-datasource is a yarn workspace
- :heavy_check_mark:  ./public/plugins/grafana-test-datasource can be built (`yarn build`), linted (`yarn lint`) and tested (`yarn test`) in that directory. You can also use `yarn workspace grafana-test-datasource run build`, `yarn workspace grafana-test-datasource run lint` and/or `yarn workspace grafana-test-datasource run test`.
- :heavy_check_mark:  ./public/plugins/grafana-test-datasource can be consumed by core builds via workspaces.
- :heavy_check_mark:  Importing frontend dependencies from ./public/plugins/grafana-test-datasource within Grafana core work as expected
- :heavy_check_mark: After removing the "old" code from its legacy location, the plugin successfully loads on its new location.
- :heavy_check_mark: Added a feature flag `runCorePluginsAsExternals` that allows to load the decoupled plugins as external, using Hashicorp plugin system (it requires to build the plugin binaries).
- :heavy_check_mark: After removing the reference to `app/features/variables/utils` (and copying the needed code to the plugin), the plugin fully loads as an external plugin.
- :heavy_check_mark: After fixing the query for live stream queries, the plugin is also fully working, running as an external plugin.
- :exclamation: verify-gen-cue now fails in CI
- :exclamation: betterer-frontend failures in CI due to using `any` types. Should be straight forward to fix.
- [ ] Testdata build, lint and tests not running in CI

**Steps I followed so far:**
<details><summary>Details</summary>
> mkdir plugins && cd plugins
> npx @grafana/create-plugin@latest
? What is going to be the name of your plugin? testdatasource
? What is the organization name of your plugin? grafana (doesn't allow empty org)
? How would you describe your plugin? Generates test data in di
fferent forms                  
? What type of plugin would you like? datasource
? Do you want a backend part of your plugin? Yes
? Do you want to add Github CI and Release workflows? Yes
? Do you want to add a Github workflow for automatically checki
ng "Grafana API compatibility" on PRs? Yes

## What's next?

Run the following commands to get started:

    * cd ./grafana-test-datasource
    * npm install to install frontend dependencies.
    * npm run dev to build (and watch) the plugin frontend code.
    * mage -v build:linux to build the plugin backend code. Rerun this command every time you edit your backend files.
    * docker-compose up to start a grafana development server. Restart this command after each time you run mage to run your new backend code.
    * Open http://localhost:3000 in your browser to create a dashboard to begin developing your plugin.

Note: We strongly recommend creating a new Git repository by running git init in ./grafana-test-datasource before continuing.

    * View create-plugin documentation at https://grafana.github.io/plugin-tools/
    * Learn more about Grafana Plugins at https://grafana.com/docs/grafana/latest/plugins/developing/development/

// skip pre-commit due to problems
> gc -nm "scaffold new testdata datasource"

// manually update package.json and plugin.json
> gc -m "rename package and plugin metadata"

> cd grafana-test-datasource
> yarn
Usage Error: The nearest package directory (/home/marcus/go/src/github.com/grafana/grafana/plugins/grafana-test-datasource) doe
sn't seem to be part of the project declared in /home/marcus/go/src/github.com/grafana/grafana.                                
                                                                                                                               
- If /home/marcus/go/src/github.com/grafana/grafana isn't intended to be a project, remove any yarn.lock and/or package.json fi
le there.                                                                                                                      
- If /home/marcus/go/src/github.com/grafana/grafana is intended to be a project, it might be that you forgot to list plugins/gr
afana-test-datasource in its workspace configuration.                                                                          
- Finally, if /home/marcus/go/src/github.com/grafana/grafana is fine and you intend plugins/grafana-test-datasource to be treat
ed as a completely separate project (not even a workspace), create an empty yarn.lock file in it.                              
                                                                                                                               
$ yarn install [--json] [--immutable] [--immutable-cache] [--check-cache] [--inline-builds] [--mode #0]  

// manually add ./plugins/* as yarn workspace in root package.json
> yarn 
...
YN0000: Done with warnings in 12s 307ms

> cd ../../

> gc -m "add ./plugins directory and any sub-directory as yarn workspace"

// manually update @grafana packages to 10.1.0-pre in package.json
> cd plugins/grafana-test-datasource && yarn 
YN0000: Done with warnings in 3s 413ms

> cd ../../
> yarn build
Success

// Backend 
// manually update module name, go version in plugins/grafana-test-datasource/go.mod
// manually update import path in plugins/grafana-test-datasource/pkg/main.go 
> cd plugins/grafana-test-datasource
> go mod tidy
> gc -m "update go module and paths"

> cd ../../ (root)

> make run 
Failure something about sub go modules

// in plugins/grafana-test-datasource get some warnings in vs code:
- on the main package 
This file is within module "plugins/grafana-test-datasource", which is not included in your workspace.
To fix this problem, you can add a go.work file that uses this directory.
See the documentation for more information on setting up your workspace:
https://github.com/golang/tools/blob/master/gopls/doc/workspace.md
- On the import of "github.com/grafana/grafana/plugins/grafana-test-datasource/pkg/plugin"
could not import github.com/grafana/grafana/plugins/grafana-test-datasource/pkg/plugin (current file is not included in a workspace module)

// Go workspaces
> go work init 
> go work use .
> go work use plugins/grafana-test-datasource
> go mod tidy 
> make run
Success

// warnings in vs code gone 

> gc -m "Add go workspaces to make build and vs code happy"

// next extract backend testdata code from core grafana to new structure 
// 1. copy go code from ./pkg/tsdb/testdatasource into ./plugins/grafana-test-datasource/pkg/plugin 
// 2. rename package name of copied files to plugin
// 3. run go tests in ./plugins/grafana-test-datasource/pkg/plugin and all internal grafana dependencies resolves
// Success 
> gc -m "copy testdata go code to new structure"

// next extract frontend testdata code from core grafana to new structure 
// 1. delete everything in ./plugins/grafana-test-datasource/src
// 2. copy from ./public/app/plugins/datasource/testdata into ./plugins/grafana-test-datasource/src
> yarn build
LOTS of errrors! Commit for now with pre-commit due to errors 
> gc -nm "copy testdata frontend code to new structure" 
// add @grafana-schema@10.1.0-pre as dependency
> yarn build 
LOTS of errrors

// trying to add dependencies, but still lots of errors
Thought I could resolve core Grafana imports using app/features... but doesn't look like it

// commiting for now
> gc -nm "trying to add missing dependencies to resolve build problems, but doesn't resolve everything"

// go mod tidy resulted in below error. 
❯ go mod tidy                                                                                                                  
go: finding module for package github.com/grafana/grafana/plugins/grafana-test-datasource/pkg/plugin                           
github.com/grafana/grafana/pkg/plugins/backendplugin/coreplugin imports                                                        
        github.com/grafana/grafana/plugins/grafana-test-datasource/pkg/plugin: no matching versions for query "latest" 

// go mod tidy currently ignores go.work files https://github.com/golang/go/issues/50750
// as a workaround had to add a replace to go.mod 

</details>

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
* So with this the test data source would still be compiled into the grafana binary (so new new process started to run plain OSS)?  
* But the plugin can be also build/deployed as an external plugin? 

We have a concept of standalone plugins (but bundled), maybe we can extend / refine that system. 
https://github.com/grafana/grafana/tree/main/plugins-bundled

File: package.json [hunk 1]
```diff
@@ -434,7 +435,8 @@
   "workspaces": {
     "packages": [
       "packages/*",
-      "plugins-bundled/internal/*"
+      "plugins-bundled/internal/*",
+      "plugins/*"
```
[the remark this answers is not in the document]
Comment (marefr) [reply package.json:439]: This is the first step/POC. Didn't want to put it there in the first step having to fix any plugins bundled related CI things first thing for now.

**Comment by andresmgot** [comment]
Summary of the changes I've pushed:

 - The plugin needs to be placed within the `public` folder, to be able to access source files like the plugin icon.
 - The asset paths need to be updated to reflect the change in the directory.
 
With that, the plugin is fully functional on its new directory. The next step would be to be able to load the plugin as "external" and skip the core version.

**Comment by marefr** [comment]
Closed in favor of #72205

**Pull request closed without merging**